[SYCL][DOC] Expand complex extension with complex support for sycl::marray (#11792)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_complex.asciidoc

While {dpcpp} has support for `std::complex` in device code, it limits the
complex interface and operations to the existing C++ standard. This proposal
defines a SYCL complex extension based on but independent of the `std::complex`
interface.

The proposed framework not only encompasses complex support for traditional use
cases but also accommodate for advanced mathematical features and data
structures.

Specifically, we propose to incorporate complex support for `sycl::marray`.
This addition will empower developers to store complex numbers seamlessly

> within arrays, opening up new possibilities for data manipulation and

Reviewer: ```suggestion
within a `sycl::marray`, opening up new possibilities for data manipulation and
```

Is this what you meant?  Developers can already create a `std::array` of `std::complex`, right?

Author: - This is what I meant with the "array", but clearly, it is not explicit enough and creates confusion.
- Correct
- I'll apply the suggestion :+1:

  /// Compares complex numbers z and w and returns true if they are different, otherwise false.
  friend constexpr bool operator!=(const complex<value_type> &z, const complex<value_type> &w);
  ///Compares complex number z and real y and returns true if they are different, otherwise false.
  friend constexpr bool operator!=(const complex<value_type> &z, value_type y);
  /// Compares real x and complex number w and returns true if they are different, otherwise false.
  friend constexpr bool operator!=(value_type x, const complex<value_type> &w);

  /// Reads a complex number from is.
  /// Not allowed in device code.
  template <class C, class T> friend std::basic_istream<C, T> &operator>>(std::basic_istream<C, T> &is, complex<value_type> &);
  /// Writes to os the complex number z in the form (real,imaginary).
  /// Not allowed in device code.
  template <class C, class T> friend std::basic_ostream<C, T> &operator<<(std::basic_ostream<C, T> &os, const complex<value_type> &);
  /// Streams the complex number z in the format "(real,imaginary)" into `sycl::stream` x and return the result.
  friend const sycl::stream &operator<<(const sycl::stream &x, const complex<value_type> &z);

> };

Author: Done

} // namespace sycl::ext::oneapi::experimental
```

=== Marray Complex Class Specialization

This proposal also introduces the specialization of the marray class to
support SYCL complex. The marray class undergoes slight modification for this
specialization, primarily involving the removal of operators that are
inapplicable. No new functions or operators are introduced to the marray class.

The marray complex specialization maintains the principles of trivial

> copyability (as seen in the Complex class description), with the

Author: Done

  friend marray<bool, NumElements> operator <=(const value_type & lhs, const marray & rhs) = delete;

  friend marray<bool, NumElements> operator >=(const marray & lhs, const marray & rhs) = delete;
  friend marray<bool, NumElements> operator >=(const marray & lhs, const value_type & rhs) = delete;
  friend marray<bool, NumElements> operator >=(const value_type & lhs, const marray & rhs) = delete;

  friend marray operator ~(const marray &v) = delete;

  friend marray<bool, NumElements> operator !(const marray &v) = delete;
};

} // namespace sycl
```

=== Scalar Mathematical operations

This proposal extends the `sycl::ext::oneapi::experimental` namespace math
functions to accept `complex<sycl::half>`, `complex<float>`, `complex<double>`
as well as the scalar types `sycl::half`, `float` and `double` for a range of
SYCL math functions.

Specifically, it adds support for `abs`, `acos`, `asin`, `atan`, `acosh`,
`asinh`, `atanh`, `arg`, `conj`, `cos`, `cosh`, `exp`, `log`, `log10`, `norm`,
`polar`, `pow`, `proj`, `sin`, `sinh`, `sqrt`, `tan`, and `tanh`.

Additionally, this extension introduces support for the `real` and `imag` free

> functions, which the real and imaginary component, respectively.

Reviewer: ```suggestion
functions, which return the real and imaginary component of a number, respectively.
```

I'm not sure, but it may be worth adding a note about what happens if the input is not a complex number here.

Author: Don't you think that the declarations of these functions shown after this block (with comments) are not enough?

```c++
/// Returns the real component of the complex number z.                                                                                                                                                                                                                                   
template <class T> constexpr T real(const complex<T> &);                                                                                                                                                                                                                                  
/// Returns the real component of the number y, treated as complex numbers with zero imaginary component.                                                                                                                                                                                 
template <class T> constexpr T real(T);                                                                                                                                                                                                                                                   
/// Returns the imaginary component of the complex number z.                                                                                                                                                                                                                              
template <class T> constexpr T imag(const complex<T> &);                                                                                                                                                                                                                                  
/// Returns the imaginary component of the number y, treated as complex numbers with zero imaginary component.                                                                                                                                                                            
template <class T> constexpr T imag(T); 
```

} // namespace sycl::ext::oneapi::experimental
```

=== Marray Complex Class Specialization

> This proposal also introduces the specialization of the marray class to

Reviewer: I'd prefer that we use `marray` and `complex` throughout here, to make it clear when we're talking about a class.

(This comment applies to the whole of this section.)

support SYCL complex. The marray class undergoes slight modification for this
specialization, primarily involving the removal of operators that are
inapplicable. No new functions or operators are introduced to the marray class.

The marray complex specialization maintains the principles of trivial
copyability (as seen in the Complex class description), with the
`is_device_copyable` type trait resolving to `std::true_type`.

The marray definition used within this proposal assumes that any operator the
`sycl::marray` class defines is only implemented if the marray's value type
also implements the operator.

For instance,
`sycl::marray<sycl::ext::oneapi::experimental::complex<T>, NumElements>` does
not implement the modulus operator since

> `sycl::ext::oneapi::experimental::complex<T>` does not support it.

Reviewer: I think you should replace this block with comments on the operators themselves.

SYCL does this by adding an "Available only when" comment, see [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_math_array_interface) for an example.

Author: I understand your suggestion, however, this is not what this "example" is trying to show.

What this example is trying to show is that the `marray` specialization only implements the operator that `complex` implements.
They will never be "Available in certain condition" since the `marray` is specialized for `complex`.
That's why the structure shown after this block only shows the operator deleted.

This might be clearer:

```
-The marray definition used within this proposal assumes that any operator the
-`sycl::marray` class defines is only implemented if the marray's value type
-also implements the operator.
+The `marray` specialization used within this proposal assumes that any operator
+the `sycl::marray` class defines is only implemented if the `marray`'s value
+type also implements the operator.
 
-For instance,
-`sycl::marray<sycl::ext::oneapi::experimental::complex<T>, NumElements>` does
-not implement the modulus operator since
-`sycl::ext::oneapi::experimental::complex<T>` does not support it.
+Because the `marray` is specialized for `complex`; as an example, the modulus
+operator is not implemented since `sycl::ext::oneapi::experimental::complex<T>`
+does not support it.
```

Reviewer: Oh, I see.

You're right; if you're using a specialization, then we shouldn't use "Available only when" for this.  I think what threw me off was your use of the word "assumes" -- I read that as something like "This proposal is relying on behavior not documented here", which made me go and look at how `sycl::marray` was defined in the SYCL spec.

I would recommend something even simpler than what you suggested.  It seems to me that something like the below would be sufficient to document the class and summarize what the class synopsis is trying to convey:

> The `marray` specialization for `complex<T>` deletes any operator that is not supported by `complex<T>`.

template <class T> complex<T> atanh(const complex<T> &);
/// Compute the hyperbolic sine of complex number x.
template <class T> complex<T> sinh(const complex<T> &);
/// Compute the hyperbolic cosine of complex number x.
template <class T> complex<T> cosh(const complex<T> &);
/// Compute the hyperbolic tangent of complex number x.
template <class T> complex<T> tanh(const complex<T> &);
/// Compute the inverse sine of complex number x.
template <class T> complex<T> asin(const complex<T> &);
/// Compute the inverse cosine of complex number x.
template <class T> complex<T> acos(const complex<T> &);
/// Compute the inverse tangent of complex number x.
template <class T> complex<T> atan(const complex<T> &);
/// Compute the sine of complex number x.
template <class T> complex<T> sin(const complex<T> &);
/// Compute the cosine of complex number x.
template <class T> complex<T> cos(const complex<T> &);
// Compute the tangent of complex number x.
template <class T> complex<T> tan(const complex<T> &);

} // namespace sycl::ext::oneapi::experimental
```

=== Element-Wise Mathematical operations

In harmony with the complex scalar operations, this proposal extends

> furthermore the `sycl::ext::oneapi::experimental`` namespace math functions

Reviewer: ```suggestion
furthermore the `sycl::ext::oneapi::experimental` namespace math functions
```

to accept `sycl::marray<complex<T>>` for a range of SYCL math functions.

Specifically, it adds support for `abs`, `acos`, `asin`, `atan`, `acosh`,
`asinh`, `atanh`, `arg`, `conj`, `cos`, `cosh`, `exp`, `log`, `log10`, `norm`,
`polar`, `pow`, `proj`, `sin`, `sinh`, `sqrt`, `tan`, and `tanh`.

Additionally, this extension introduces support for the `real` and `imag` free
functions, which return marrays of scalar values representing the real and

> imaginary components, respectively.

Reviewer: Same comment as above.

Author: Don't you think that the declarations of these functions shown after this block (with comments) are not enough?

```c++
/// Returns an marray of real components from the marray x.                                                                                                                                                                                                                               
template <typename T, std::size_t NumElements>                                                                                                                                                                                                                                            
sycl::marray<T, NumElements> real(const marray<complex<T>, NumElements> &x);                                                                                                                                                                                                              
/// Returns an marray of imaginary components from the marray x.                                                                                                                                                                                                                          
template <typename T, std::size_t NumElements>                                                                                                                                                                                                                                            
sycl::marray<T, NumElements> imag(const marray<complex<T>, NumElements> &x);
```

Reviewer: We don't usually specify behavior in comments like this, so I don't think a reader would think to check them if they were looking for a description of what the functions do.  They're also not a complete description, because they don't explain which values of `T` are supported.

I can see that the comment style was being used by this specification before you modified it, so I'm happy if you want to leave the formatting updates for a later PR.

Author: In the `Complex Class` section, it is specified that "The `T` template parameter must be one of the types float, double, or sycl::half."

A possible solution could be to add a note that links to this part of the `Complex Class`.
What do you think?

Reviewer: I'm not talking about the overloads that accept a `complex<T>`, but the cases where you don't accept a complex number:

```c++
template <class T> constexpr T real(T);
template <class T> constexpr T imag(T);
...
// and so on
```

Presumably these are intended to support the same types as the ISO C++ version, but I don't see that said anywhere.  Also note that ISO C++ doesn't define these functions this way, but rather as `real(float)`, `real(double)` and `real(long double)`.

Reviewer: We shouldn't concern ourselves too much with what cppreference says, because it's just an online reference guide.  It's more important in my opinion to align with the ISO C++ specification, which defines these additional overloads [here](https://eel.is/c++draft/cmplx.over).

Rather than duplicate all of this text, though, the simplest thing to do here would probably just be to add a note explaining that these functions aren't really new, and that their behavior matches ISO C++.

But as you say, this isn't really related to `marray`.  I'm happy for it to be addressed in a later PR.

Author: I'll create a PR on top of this branch and will ping you to review the new note added to the `real` and `imag` functions.

I will let you `Resolve the conversation` as a sign of consensus and closure on that topic.
